File: Ch06/06_04/shutdown.py
import logging
from os import remove

logging.basicConfig(
    format=(
        '%(asctime)s - %(name)s:%(levelname)s - %(filename)s:%(lineno)d'
        '- %(message)s'
    ),
    level=logging.INFO,
)
logger = logging.getLogger('shutdown')


def shutdown_server(pid_file):
    try:
        with open(pid_file) as fp:
            pid = int(fp.read())
    except (OSError, ValueError) as err:
        logger.error('%r: cannot get pid - %s', pid_file, err)
        raise

    logger.info('shutting down %s', pid)
    # TODO: Actual shutdown

    try:
        remove(pid_file)
    except OSError as err:
        logger.warning('cannot delete %r - %s', pid_file, err)


if __name__ == '__main__':
    shutdown_server('Ch06/06_04/app.pid')

File: Ch06/challenge/shutdown.go
package main

import (
	"fmt"
	"log"
	"os"
)

var (
	logger = log.New(log.Writer(), "[shutdown ]", log.LstdFlags|log.Lshortfile)
)

func shutdownServer(pidFile string) error {
	file, err := os.Open(pidFile)
	if err != nil {
		logger.Printf("ERROR: can't open %q", pidFile)
		return err
	}
	defer file.Close()

	var pid int
	if _, err := fmt.Fscanf(file, "%d", &pid); err != nil {
		logger.Printf("ERROR: %q: bad pid - %s", pidFile, err)
		return err

	}

	logger.Printf("INFO: shutting down %d", pid)
	// TODO: Actual shutdown

	if err := os.Remove(pidFile); err != nil {
		log.Printf("WARNING: can't delete %q - %s", pidFile, err)
	}

	return nil
}

func main() {
	shutdownServer("app.pid")
}
